contracts/PushStaking/PushStaking.sol

```diff
@@ -149,31 +150,40 @@ contract PushStaking is Initializable, PushStakingStorage {
     }
 
     function decreaseWalletShare(
-        address _walletAddress,
-        GenericTypes.Percentage memory _percentage
-    )
-        external
-        onlyGovernance
-    {
-        uint256 sharesToBeRemoved = walletShareInfo[_walletAddress].walletShare;
-        removeWalletShare(_walletAddress);
-        addWalletShare(_walletAddress, _percentage);
-        emit SharesDecreased(_walletAddress, sharesToBeRemoved, walletShareInfo[_walletAddress].walletShare);
-    }
+            address _walletAddress,
+            GenericTypes.Percentage memory _percentage
+        )
+            external
+            onlyGovernance
+        {
+            if(_walletAddress == address(0) || _walletAddress == FOUNDATION) {
+               revert Errors.InvalidArgument_WrongAddress(_walletAddress);
+            }
+
+            uint currentEpoch = lastEpochRelative(genesisEpoch,block.number);
+
+            uint256 currentShares = walletShareInfo[_walletAddress].walletShare;
+            uint256 sharesToBeAllocated = BaseHelper.calcPercentage(WALLET_TOTAL_SHARES, _percentage);
 
-    /*
-     *Reward Calculation for a Given Wallet
-     * 1. calculateWalletRewards(address wallet)
-     * 2. public helper function
-     * 3. Helps in calculating rewards for a specific wallet based on
-       * a. Their Wallet Share 
-        * b. Total Wallet Shares
-        * c. Total Rewards available in WALLET_TOTAL_SHARES
-     * 4. Once calculated rewards for a sepcific wallet can be updated in WalletToRewards mapping
-     * 5. Reward can be calculated for wallets similar they are calculated for Token holders in a specific epoch.
-        * Reward for a Given Wallet X = ( Wallet Share of X / WALLET_TOTAL_SHARES) * WALLET_FEE_POOL
-    */
-    //TODO logic yet to be finalized
+            if(sharesToBeAllocated >= currentShares){
+                revert Errors.InvalidArg_MoreThanExpected(currentShares, sharesToBeAllocated);
+            }
+            uint256 sharesToBeRemoved = currentShares - sharesToBeAllocated;
+            walletShareInfo[_walletAddress].walletShare = sharesToBeAllocated;
```

Comment:
Is there any reason why _adjustWalletAndTotalStake fn is not called for both foundation and the wallet instead of directly updating the state variables?
@Zartaj0

Reply:
Ideally  `_adjustWalletAndTotalStake` should be called, but I've left it as of now because we can't pass the `sharesToAdd` and `sharesToRemove` values as it will increase the total share variables. And if we pass zero values for them, it means we're not effectively updating variables in `_adjustWalletAndTotalStake` or the `_setupEpochsRewardAndSharesForWallets`  functon, other than epochRewards mapping of course. And it also significantly increases the gas cost. 

The current implementation also has a problem of not updating the `stakedBlock`, as it only gets updated when `sharesToAdd` is non-zero.

Reply:
Cool, it's okay for now then
It might trouble at the time of merging wallets and token staking functions.